src/txmempool.cpp

```diff
@@ -854,7 +854,8 @@ void CTxMemPool::PrioritiseTransaction(const uint256& hash, const CAmount& nFeeD
 
 void CTxMemPool::ApplyDelta(const uint256 hash, CAmount &nFeeDelta) const
 {
-    LOCK(cs);
+    AssertLockHeld(cs);
```

Comment:
https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#assert-capability-and-assert-shared-capability:
> These are attributes on a function or method that does a run-time test to see whether the calling thread holds the given capability.

How this could disable clang's compile time checks?

> ... if `ApplyDelta` lost its annotation, it could then be called from somewhere without the lock held and without raising errors if it called other functions expecting cs to be held...

Why `ApplyDelta` could lost its annotation?
Locking a recursive mutex in any site where we think it should be locked seems worse than using a non-recursive mutex with annotated functions and data members, no?

Reply:
> How this could disable clang's compile time checks?

They're attributes for a function that does a runtime test and unwinds the stack, ie a function like `f(lock) { if (!lock) throw; }`, at which point it's safe to assume that the lock is held after the function returns, so the compile time checks are no longer necessary, so disabling the compile time checks is the whole point. (That's true for `AssertLockHeldInternal` when `DEBUG_LOCK_ORDER` is defined, but not when it's not defined)

You can see this if you change `LOCK(cs)` in `CTxMemPool::isSpent()` to `AssertLockHeld(cs)`. Suddenly dispite the `GUARDED_BY(cs)`, access to `mapNextTx` is allowed at compile time, and you'll only get an error at runtime, and then only if you compile with `DEBUG_LOCK_ORDER`.

This feature is needed by net_processing.cpp in `MaybeSetPeerAsAnnouncingHeaderAndIDs` where it passes a lambda to `connman.ForNode` where the lambda needs to know a lock is held to access `lNodesAnnouncingHeaderAndIDs`, but `ForNode` is too general to be able to assure the lambda that the lock is held.

> Why `ApplyDelta` could lost its annotation?

Someone makes a typo; it's not especially likely, but with a `GUARDED_BY` and without `AssertLockHeld` the missing annotation would be caught at compile time.

> Locking a recursive mutex in any site where we think it should be locked seems worse than using a non-recursive mutex with annotated functions and data members, no?

No, no -- your locking changes make sense, just saying that the `AssertLockHeldInternal` annotation isn't helpful here. (There's lots of other places where it's not helpful too)

Reply:
[Updated](https://github.com/bitcoin/bitcoin/pull/19647#issuecomment-668106705).

Reply:
> Wouldn't it be better just to rely on the compile time checks?

I find this scary. Those compile time checks do not look mature enough to me - see how the behavior changes depending on the order of the attributes: https://github.com/bitcoin/bitcoin/pull/19668#discussion_r467045878. Also, they do not work with gcc and even with clang produce only a compile time warnings which could be missed if -Werror is not used.

Reply:
> `AssertLockHeld` is annotated with `ASSERT_EXCLUSIVE_LOCK` which disables clang's compile time checks that the lock was acquired beforehand.

It is fixed in #19668.